File: algorithms/lua/huffman.lua
local function build_freq(text)
    local freq = {}
    for i = 1, #text do
        local c = text:sub(i,i)
        freq[c] = (freq[c] or 0) + 1
    end
    return freq
end

local function build_tree(freq)
    local nodes = {}
    for ch, f in pairs(freq) do
        table.insert(nodes, {freq=f, ch=ch, left=nil, right=nil})
    end

    local function pop_min()
        local mi, mv = 1, nodes[1].freq
        for i = 2, #nodes do
            if nodes[i].freq < mv then mi = i; mv = nodes[i].freq end
        end
        local n = nodes[mi]
        table.remove(nodes, mi)
        return n
    end

    while #nodes > 1 do
        local l = pop_min()
        local r = pop_min()
        table.insert(nodes, {freq=l.freq+r.freq, ch=nil, left=l, right=r})
    end
    return nodes[1]
end

local function build_codes(node, depth, codes)
    if not node then return end
    if not node.left and not node.right then
        codes[node.ch] = depth > 0 and depth or 1
        return
    end
    build_codes(node.left,  depth+1, codes)
    build_codes(node.right, depth+1, codes)
end

local function main()
    local text  = io.read("*l")
    local n     = #text
    local freq  = build_freq(text)

    local start = os.clock()

    local root  = build_tree(freq)
    local codes = {}
    build_codes(root, 0, codes)

    local unique     = 0
    local total_bits = 0
    for ch, f in pairs(freq) do
        unique = unique + 1
        total_bits = total_bits + f * (codes[ch] or 1)
    end

    local ms    = (os.clock() - start) * 1000
    local ratio = total_bits / (8.0 * n)

    print(unique)
    print(total_bits)
    print(string.format("%.2f", ratio))
    io.stderr:write(string.format("%.3f\n", ms))
end

main()

File: algorithms/go/huffman.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"time"
)

type Node struct {
	freq        int
	ch          byte
	left, right *Node
}

type PQ []*Node

func (pq PQ) Len() int            { return len(pq) }
func (pq PQ) Less(i, j int) bool  { return pq[i].freq < pq[j].freq }
func (pq PQ) Swap(i, j int)       { pq[i], pq[j] = pq[j], pq[i] }
func (pq *PQ) Push(x interface{}) { *pq = append(*pq, x.(*Node)) }
func (pq *PQ) Pop() interface{} {
	old := *pq; n := len(old)
	x := old[n-1]; *pq = old[:n-1]
	return x
}

var codes [256]int

func buildCodes(n *Node, depth int) {
	if n == nil { return }
	if n.left == nil && n.right == nil {
		if depth == 0 { depth = 1 }
		codes[n.ch] = depth
		return
	}
	buildCodes(n.left,  depth+1)
	buildCodes(n.right, depth+1)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var text string
	fmt.Fscan(reader, &text)
	n := len(text)

	var freq [256]int
	for i := 0; i < n; i++ { freq[text[i]]++ }

	start := time.Now()

	pq := &PQ{}
	heap.Init(pq)
	unique := 0
	for ch, f := range freq {
		if f > 0 {
			heap.Push(pq, &Node{freq: f, ch: byte(ch)})
			unique++
		}
	}

	for pq.Len() > 1 {
		l := heap.Pop(pq).(*Node)
		r := heap.Pop(pq).(*Node)
		heap.Push(pq, &Node{freq: l.freq + r.freq, left: l, right: r})
	}

	if pq.Len() > 0 {
		buildCodes(heap.Pop(pq).(*Node), 0)
	}

	totalBits := 0
	for ch, f := range freq {
		if f > 0 { totalBits += f * codes[ch] }
	}

	ms    := float64(time.Since(start).Nanoseconds()) / 1e6
	ratio := float64(totalBits) / (8.0 * float64(n))

	fmt.Fprintln(writer, unique)
	fmt.Fprintln(writer, totalBits)
	fmt.Fprintf(writer, "%.2f\n", ratio)
	fmt.Fprintf(os.Stderr, "%.3f\n", ms)
}